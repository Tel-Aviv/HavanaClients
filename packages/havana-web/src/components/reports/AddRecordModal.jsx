// @flow
import React from 'react';
import moment from 'moment';
import { Popconfirm, Modal, Form, Icon, Button, 
    Typography , Input, Row, Col } from 'antd';
const { Title } = Typography;    
import { useTranslation } from "react-i18next";   

import CustomTimePicker from '../CustomTimePicker'

const AddRecordModal = (props) => {

    const { t } = useTranslation();

    const { getFieldDecorator } = props.form;

    const visible = props.visible;
    const onCancel = props.onCancel;
    const _addRecord = props.onAddRecord;

    const _onSubmit = e => {

        e.preventDefault();

        const fieldsValue = props.form.getFieldsValue();
        console.log(fieldsValue);
        if( moment(fieldsValue.entryTime).format("HH:mm") >= moment(fieldsValue.exitTime).format("HH:mm") ) {
            props.form.setFields({
                entryTime: {
                value: fieldsValue.entryTime,
                errors: [new Error(t('exit_before_entry'))],
              },
            });
            return;
        }

        props.form.validateFields( (err, values) => {

            if (err) {
                return;
            }

            console.log(values);    
            
            const _values = {
                inTime: values["entryTime"],
                outTime: values["exitTime"],
                note: values["notes"]
            }   
            
            if( _addRecord )
                _addRecord(_values);
        })
    }    

    return (
        <Modal visible={visible}
              closable={false}
              className='rtl'
              footer={[
                <Button key='cancel' onClick={onCancel}>
                    {t('cancel')}
                </Button>,
                <Button key='submit' onClick={_onSubmit} 
                        type="primary" htmlType="submit">
                    {t('add_record')}
                </Button>
              ]}>
            <Title level={3} className='rtl'
                style={{
                    marginTop: '12px'
                }}>
                {t('add_record')}{ t('to_day')} 
                { 
                    props.record ?  
                    props.record.rdate : null
                }
            </Title> 

            <Form layout="vertical"
                    size='small'
                    onSubmit={_onSubmit}>
                <Form.Item label={t('in')}>
                    {getFieldDecorator('entryTime', {
                        rules: [{ 
                                type: 'object', 
                                required: true, 
                                message: t('add_entry_error') 
                                }],
                    })(
                        <CustomTimePicker />
                    )}
                </Form.Item>
                <Form.Item label={t('out')}>
                    {getFieldDecorator('exitTime', {
                        rules: [{ 
                                type: 'object', 
                                required: true, 
                                message: t('add_exit_error') 
                                }],                
                    })(
                        <CustomTimePicker />
                    )}
                </Form.Item>
                <Form.Item label={t('notes')}>
                    {getFieldDecorator('notes', {
                        rules: [{ required: true, message: t('add_notes_error') }],
                    })(
                        <Input />,
                    )}
                </Form.Item>                
            </Form>
        </Modal>        
    )
}

export default Form.create()(AddRecordModal)