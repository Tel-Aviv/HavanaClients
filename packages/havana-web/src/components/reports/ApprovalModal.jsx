import React from 'react';
import { Modal, Steps, Button, Typography } from 'antd';
const { Title } = Typography;

const { Step } = Steps;
import { useTranslation } from "react-i18next";

import ExtraHoursDiff from './approval/ExtraHoursDiff';
import FinalApproval  from './approval/FinalApproval';

const ApprovalModal = ({visible, onCancel, onOk, extraHours}) => {

    const [currentStep, setCurrentStep] = React.useState(0);

    const { t } = useTranslation();

    const steps = [{
        title: t('extra_hours'),
        content: <ExtraHoursDiff extraHours={extraHours} />
    }, {
        title: t('approve'),
        content: <FinalApproval />
    }]

    const next = () => {
        setCurrentStep(currentStep + 1);
      };
    
    const prev = () => {
        setCurrentStep(currentStep - 1);
    };

    return <Modal visible={visible}
                    className='rtl'
                    closable={true}
                    footer={null}
                    onCancel={onCancel}
                    onOk={onOk}>
        <Title level={3} className='rtl'>
            {t('approve')}
        </Title>
        <Steps current={currentStep}
            style={{
                direction: 'rtl'
            }}>
            {
                steps.map( (step, index) => 
                    <Step key={index} title={step.title} />
                )
            }
        </Steps>
        <div className="steps-content">
        {
            steps[currentStep].content
        }
        </div>
        <div className="steps-action">
            {currentStep < steps.length - 1 && (
                <Button type="primary" onClick={() => next()}>
                    {t('next')}
                </Button>
            )}
            {currentStep === steps.length - 1 && (
                <Button type="primary">
                    Done
                </Button>
            )}            
            {currentStep > 0 && (
                <Button style={{ margin: '0 8px' }} onClick={() => prev()}>
                    {t('previous')}
                </Button>
            )}
        </div>
    </Modal>
    
}

export default ApprovalModal;