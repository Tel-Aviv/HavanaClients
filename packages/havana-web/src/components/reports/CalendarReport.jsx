import React, { useEffect, useState } from 'react';
import { Calendar, Row, Col, Tag,
    Modal, Tooltip } from 'antd';
import moment from 'moment';
import { useTranslation } from "react-i18next";

const DailyTable = React.lazy( () => import('./DailyTable') );

const format = 'HH:mm';

const CalendarReport = (props) => {

    const [originalData, setOriginalData] = useState([]);
    const [firstLevelData, setFirstLevelData] = useState([]);
    const [dailyReportVisible, setDailyReportVisible] = useState(false);
    const [secondLevelData, setSecondLevelData] = useState([]);
    const [selectedDate, setSelectedDate] = useState();

    const { t } = useTranslation();

    useEffect( () => {

        if( !props.dataSource || props.dataSource.length === 0 )
            return;

        const _originalData = props.dataSource.map( record =>  {

            return { 
                ...record, 
                rdate : moment(record.rdate)
            }
          })

        setFirstLevelData(
            _originalData.filter( 
                item => item.stripId === 1)
        );

        setOriginalData(_originalData);

    }, [props.dataSource])

    const onSelect = value => {

        // No edits for the future
        if( value.isAfter(moment()) )
            return;

        const day = value.date();
        setSecondLevelData(getSecondLevelData(day));
        setDailyReportVisible(true);
        setSelectedDate(value.format('DD/MM/YYYY'))
    }

    const onDailyReportClosed = () => {
        setSelectedDate(null);
        setDailyReportVisible(false);
    }

    const dateFullCellRender = (date) => {
        const originalItem = firstLevelData.find( item => item.rdate.isSame(date, 'day') )

        if( !originalItem )  
            return null;

        return <div className='ant-picker-cell-inner ant-picker-calendar-date'>
            <div className='ant-picker-calendar-date-value'>
                <span className='calendar-date'>
                    {date.date()}
                </span>
            </div>
            <div className='ant-picker-calendar-date-content'>
        <Row>
            <Col>
            {
                ( !isInPast(originalItem) ) ?
                    <ul className='calendar-events'>
                        <li style={{
                            margin: '3px'
                        }}>
                            { t('required') + ': ' + originalItem.requiredHours }
                        </li>
                        <li style={{
                            margin: '3px'
                        }}>
                            { t('accepted') + ': ' + originalItem.acceptedHours }
                        </li>
                        <li style={{
                            float: 'right'
                        }}>
                            { originalItem.systemNotes ? 
                                <Tag color='magenta'
                                    style={{
                                        marginRight: '0',
                                        width: '100%',
                                        textAlign: 'start'
                                    }}>
                                    {originalItem.systemNotes}
                                </Tag> : null
                            }
                        </li>
                    </ul>
                : null
            }
            </Col>
        </Row>
        </div>
        </div>   
    }

    const getSecondLevelData = (day) => {

        const dailyData =  originalData.filter( item =>
            parseInt(item.day) === day
        )
        
        return ( dailyData.length === 1 ) ?
            dailyData :
            dailyData.filter( item => parseInt(item.stripId) > 1 )
    }

    const isInPast = (record) => {
        return moment(record.rdate, 'DD/MM/YYYY').isAfter(moment());
    }

    const saveRecord = async (record, inouts) => {
    
        try {

          const newData = [...originalData];
          const index = newData.findIndex(item => record.key === item.key);
          if (index > -1) {
            const item = newData[index];
            let newItem = {
              ...item,
              ...record,
              rdate: moment(item.rdate, 'DD/MM/YYYY').startOf('day').format('DD/MM/YYYY')
            }
            newItem.total = moment.utc(moment(newItem.exit, format).diff(moment(newItem.entry, format))).format(format)
            newItem.valid = true;
            
            newData.splice(index, 1, newItem);
            props.onChange && props.onChange(newItem, inouts);  
            setOriginalData(newData)
          }

        } catch( errorInfo ) {
          console.error(errorInfo)
        }

      }

      const replaceRecord = (key, newItem) => {
        const newData = [...originalData];
        const index = newData.findIndex(item => key === item.key);
        if (index > -1) {

          const item = newData[index];
          let replacedItem = {
            ...item,
            inTime: newItem.inTime, 
            outTime: newItem.outTime, 
            rdate: moment(item.rdate, 'DD/MM/YYYY').startOf('day').format('YYYY-MM-DD'),
            reportCode: newItem.reportCode, 
            userNotes: newItem.userNotes,
            isFullDay: true
          }

          replacedItem.valid = true;
          newData.splice(index, 1, replacedItem);
          setOriginalData(newData);
          //setRecordToAdd(null);

          props.onChange && props.onChange(replacedItem, null);
        }
      }

    return (
        <>
            <Modal title={selectedDate}
                visible={dailyReportVisible}
                destroyOnClose='true'
                footer={null}
                width='64%'
                closable={true} 
                className='rtl'
                onCancel={ () => onDailyReportClosed() }
                onOk={ () => onDailyReportClosed() }>
                <DailyTable 
                    dataSource={secondLevelData}
                    reportCodes={props.reportCodes}
                    daysOff={props.daysOff}
                    manualUpdates={props.manualUpdates}
                    editable={true}
                    onSave={saveRecord}
                    onReplace={replaceRecord}/>
            </Modal>

            {
                originalData.length > 0 ?
                <Calendar
                    style={{
                        margin: '12px'
                    }}
                    mode='month'
                    onSelect={onSelect}
                    dateFullCellRender={dateFullCellRender}
                    value={new moment(originalData[0].rdate)}
                    headerRender={ () => null }
                /> : null
            }
        </>

    )
}

export default CalendarReport;