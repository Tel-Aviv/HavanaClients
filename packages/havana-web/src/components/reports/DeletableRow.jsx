import React, { useContext } from 'react';
import { ReportContext } from "./TableContext";     

const DeletableRow = (props) => {

    const context = useContext(ReportContext);
    const record = context.tableData.find( item =>
       item.key === props["data-row-key"]
    )

    const trClassName = record && record.isDeleted ? 'strikeout' : '';

    return <tr className={trClassName}>
        {
            props.children
        }
    </tr>
}

export default DeletableRow;