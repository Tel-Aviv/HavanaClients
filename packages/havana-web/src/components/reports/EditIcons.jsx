// @flow
import React from 'react'
import { Popconfirm, Icon, Tooltip } from 'antd';
import { CheckCircleTwoTone,
  CloseCircleTwoTone,
  EditTwoTone } 
from '@ant-design/icons';
import { useTranslation } from "react-i18next";

import { ReportContext } from "./TableContext";

const iconStyle = {
    margin: 8,
    fontSize: '100%'
}

export default ({record, editing, disable, edit, save, cancel}) => {

    const {t} = useTranslation();

    return editing ? (
      <span>
        <Popconfirm title={t('cancel_edit')} onConfirm={() => cancel()}
                    className='rtl'>
          <CloseCircleTwoTone twoToneColor="#eb2f96" style={iconStyle} />
        </Popconfirm>
            <CheckCircleTwoTone twoToneColor="#52c41a"
              onClick={() => save(record.key)}
              style={iconStyle} />
      </span>
    ) : (
          disable ?
            (<EditTwoTone style={iconStyle} />) :
            (<Tooltip title={t('edit_record')}>
                <EditTwoTone
                      onClick={() => edit(record)} 
                      style={iconStyle} />
              </Tooltip>) 
    )
  }
