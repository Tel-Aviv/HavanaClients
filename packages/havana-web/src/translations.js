const translations = {
    en: {
        translation: {
            "Switch": "English",
            "product_name": "Attendance system",
            "to": " to day ",
            "cancel": "Cancel",
            "sure": "Are you sure?",
            "title": "Hours Report",
            "home": "Home",
            "settings": "Settings",
            "signature": "Signature",
            "stamp": "Stamp",
            "setting_title": "Your Pesrsonal Info",
            "details": "Info",
            "notifications": "Notifications",
            "plain": "Plain",
            "calendar": "Calendar",
            "yearly": "Yearly",
            "no_entry": "no entry",
            "no_exit": "no exit",
            "submit": "Submit",
            "approve": "Approve",
            "for_month": "for month",
            "miising_in": "missing entry",
            "missing_out": "missing exit",
            "in": "In",
            "out": "Out",
            "notes": "Notes",
            "notes_for_report": "Notes for Report",
            "missing_notes": "Please provide notes",
            "not_approved": "This report is not approves yet",
            "was_approved": "This report was approved at ",
            "not_submitted": "This report was not submitted yet",
            "print": "Print",
            "continue": "Continue",
            "upload": "Upload",
            "abs_docs": "Absense Docs",
            "save": "Save",
            "save_tooltip": "Save interim",
            "saved": "Your changes were saved",
            "hours": "hours",
            "months": "months",
            "send_to_approval": "Are you sure?",
            "validate": "validate",
            "close": "close",
            "invalid_items": "Invalid items",
            "validate_report": "Validate report",
            "report_validated": "Report is valid",
            "print_report": "Print Report",
            "add_record": "Add record",
            "edit_record": "Edit hourly",
            "add_entry_error": "Entry time is mandatory",
            "add_exit_error": "Exit time is mandatory",            
            "add_notes_error": "Please provide your notes",
            "move_to": "Forward",
            "exit_before_entry": "Exit before entry?",
            "manual_tag": "Has been edited manually",
            "summary": "Summary attendance for ",
            "printed_when": "Printed on ",
            "approved_when": "Approved on ",
            "approvals_list": "Reports assigned to you",
            "pending_reports": "Pending",
            "approved_reports": "Approved",
            "m1": "January",
            "m2": "Ferbuary",
            "m3": "March",
            "m4": "April",
            "m5": "May",
            "m6": "Juny",
            "m7": "July",
            "m8": "August",
            "m9": "September",
            "m10": "October",
            "m11": "November",
            "m12": "December",
            "reports_to": "Reports to",
            "no_direct_manager": "Could not find the direct manager. Please select one from the list nearby",
            "reject": "Reject",
            "approve_tooltip" : "Accept with possible comments",
            "reject_tooltip": "Reject and provide the reason",
            "forward_tooltip": "Move the approval to the next management level",
            "rejected_note": "This report was rejected by the manager with note: ",
            "rejected_reports": "Rejected",
            "approval_status": "Pending appoval",
            "resubmitted_status": "Re-submitted",
            "resubmitted_header": "This report was resubmitted",
            "report_code": "Report Type",
            "no_managers": "No managers were defined for this user",
            "support_page": "Support Page for Report",
            "expand": "Expand",
            "collapse": "Collapse",
            "refresh": "Refresh",
            "cancel_edit": "Sure to discard changes?",
            "why_add_record": "Free text - the essence of the activity",
            "report_full_day": "Report whole day",
            "job_description": "Job Description",
            "job_description_tooltip": "This text will be attached to a daily report",
            "job_description_error": "What do you did on this day?"
        }
    },
    he: {
        translation: {
            "Switch": "Hebrew",
            "product_name": "מערכת נוכחות",
            "to_day": " ליום ",
            "cancel": "ביטול",
            "sure": "האם ברצונך לבטל את השינויים?",
            "title": "דוח שעות",
            "home": "דף ראשי",
            "settings": "הגדרות אישיות",
            "signature": "חתימה",
            "stamp": "חותמת",
            "setting_title": "הגדרות אישיות",
            "details": "פרטים",
            "notifications": "דוחות לטיפולך",
            "plain": "טבלה",
            "calendar": "לוח חודשי",
            "yearly": "לוח שנתי",
            "no_entry": "כניסה חסרה",
            "no_exit": "יציאה חסרה",
            "submit": "העבר לאישור",
            "approve": "אישור",
            "for_month": " לחודש",
            "miising_in": "חסרה כניסה",
            "missing_out": "חסרה יציאה",
            "in": "כניסה",
            "out": "יציאה",
            "notes": "הערות",
            "notes_for_report": "הערות לדוח",
            "missing_notes": "אנא הזן הערות!",
            "not_approved": "דוח זה ממתין לאישור",
            "was_approved": "הדוח הזה אושר בתקריך",
            "not_submitted": "דוח זה לא נשלח לאישור",  
            "print": "הדפס",
            "continue": "המשך",
            "upload": "העלאה",
            "abs_docs": "מסמכי היעדרות",
            "save": "שמור",
            "save_tooltip": "שמור זמנית",
            "saved": "השינויים נשמרו",
            "hours": "שעות",
            "months": "חודשים",
            "send_to_approval": "הדוח שלך יישלח לאישור על ידי מנהל הישיר. לא ייתאפשרו עדכונים בדוח לפני האישור. לשלוח?",
            "validate": "בדוק",
            "close": "סגור",
            "invalid_items": "דיווחים לא תקינים",
            "validate_report": "אמת נתונים",
            "report_validated": "נתוני הדוח תקינים",
            "print_report": "הדפסת דוח",
            "add_record": "הוספת דיווח",
            "edit_record": "דיווח שעתי",
            "add_entry_error": "שכחת זמן כניסה?",
            "add_exit_error": "שכחת זמן יציאה?",            
            "add_notes_error": "בבקשה להוסיף הערות",
            "move_to": "העבר",
            "exit_before_entry": "יציאה לפני הכניסה?",
            "manual_tag": "נערך באופן ידני",
            "summary": "גיליון נוכחות מסכם לחודש",
            "printed_when": " הופק בתאריך ",
            "approved_when": "אושר בתאריך",
            "approvals_list": "דוחות בטיפולך",
            "pending_reports": "ממתינים",
            "approved_reports": "אושרו",
            "m1": "ינואר",
            "m2": "פברואר",
            "m3": "מרץ",
            "m4": "אפריל",
            "m5": "מאי",
            "m6": "יוני",
            "m7": "יולי",
            "m8": "אוגוסט",
            "m9": "ספטמבר",
            "m10": "אוקטובר",
            "m11": "נובמבר",
            "m12": "דצמבר",
            "reports_to": "מדווח ל",
            "no_direct_manager": "לא נמצא מנהל ישיר לאישור דוחות. נא בחר מנהל ישיר מהרשימה ליד הכפתור.",
            "reject": "דחה",
            "approve_tooltip" : "לאשר עם הערות אפשריות",
            "reject_tooltip": "דחה את הדוח והסבר מדוע",
            "forward_tooltip" : "העבר את האישור לדרג ניהול הבא",
            "rejected_note": "הדוח הזה נדחה על ידי המנהל עם ההערה: ",
            "rejected_reports": "נדחו",
            "approval_status": "ממתין לאישור",
            "resubmitted_status": "נשלח מחדש",
            "resubmitted_header": "דוח זה נשלח לאישור מחדש לאחר טיפול בהערת מנהל",
            "report_code": "קוד דיווח",
            "no_managers": "לא הוגדרו מנהלים עבור משתמש זה",
            "support_page": "דף תמיכה לדוח",
            "expand": "הרחב",
            "collapse": "צמצם",
            "refresh": "רענן",
            "cancel_edit": "האם ברצונך לבטל את השינויים ?",
            "why_add_record": "טקסט חופשי - מהות הפעילות",
            "report_full_day": "דיווח יום מלא",
            "job_description": "פירוט העבודה",
            "job_description_tooltip": "הפירוט ייצורף לדוח יומי",
            "job_description_error": "מה עשית ביום זה?"    
        }
    }
}    

export default translations;
