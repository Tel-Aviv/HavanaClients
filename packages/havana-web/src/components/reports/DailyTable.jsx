import React, { useEffect, useState, Suspense } from 'react';
import { useDispatch } from 'react-redux';
import { ADD_ITEM, UPDATE_ITEM, DELETE_ITEM } from '../../redux/actionTypes';
import { Table, Popconfirm, Modal, Form, Icon,
    Tag, Row, Col, Tooltip, Typography } from 'antd';
import Ellipsis from 'ant-design-pro/lib/Ellipsis';
import { PlusCircleTwoTone, 
    MinusCircleTwoTone,
    TagOutlined } 
    from '@ant-design/icons';
import { useTranslation } from "react-i18next";
import { ReportContext } from "./TableContext";       
import moment from 'moment'; 

import EditIcons from './EditIcons';
import EditableCell from './EditableCell';
import AddRecordModal from './AddRecordModal';
const FullDayReport = React.lazy( () => import('./FullDayReport') );
//import FullDayReport from './FullDayReport';

const format = 'HH:mm';

const DailyTable = (props) => {

    const [form] = Form.useForm()

    const [tableData, setTableData] = useState([]);
    const [reportCodes, setReportCodes] = useState([]);
    const [manualUpdates, setManualUpdates] = useState([]);
    const [editingKey, setEditingKey] = useState('')

    // Full Day Report
    const [fullDayReportVisible, setFullDayReportVisible] = useState(false); 

    // States for adding new record
    const [addModalVisible, setAddModalVisible] = useState(false)
    const [recordToAdd, setRecordToAdd] = useState();

    const { t } = useTranslation();

    const dispatch = useDispatch();
    const action_ItemAdded = (item, index) => ({
        type: ADD_ITEM,
        addIndex: index,
        addItem: item
      });  
    
      const action_ItemDeleted = (item, index) => ({
        type: DELETE_ITEM,
        deleteIndex: index,
        deletedItem: item
      })

      const action_ItemUpdated = (item, index) => ({
        type: UPDATE_ITEM,
        item
      })

    useEffect( () => {

        const _originalData = props.dataSource.map( record =>  {

            const _isRowEditable = isRowEditable(record);

            return { 
                ...record, 
                requireChange : _isRowEditable, 
                valid : _isRowEditable ? false : true
            }
          })

          setTableData(_originalData)
       
    }, [props.dataSource])

    useEffect( () => {

        if( props.reportCodes )
            setReportCodes(props.reportCodes);
    
    }, [props.reportCodes])    
        
    useEffect( () => {

        setManualUpdates(props.manualUpdates);
    
    }, [props.manualUpdates])

    const isRowEditing = record => {
        return record.key === editingKey
    }

    const handleAddRow = (record) => {
        setRecordToAdd(record);
        setAddModalVisible(true);
    }

    const manuallyEditedTag = ( isEditedManually ) => {

        return isEditedManually ?
                <Tooltip title={t('manual_tag')}>
                  <Tag color='magenta'>
                        <TagOutlined />
                  </Tag> 
                </Tooltip>: null
    }

    const hasSytemNotes = (record) => {
        return record.systemNotes && record.systemNotes.startsWith('*');
    }

    const isTotalled = (item) => {
        const tokens = item.acceptedHours.split(':');
        const hours = parseInt(tokens[0]);
        const minutes = parseInt(tokens[1]);
        return item.total != '0:00';
    }

    const isWorkingDay = (item) => {
    
        const itemDate = moment(item.rdate);
    
        const index = props.daysOff.findIndex( dayOff => 
             dayOff.getDate() === itemDate.date() &&
             dayOff.getMonth() === itemDate.month() &&
             dayOff.getFullYear() === itemDate.year()
        );
    
        return index !== -1 
          ? false:   
          !(itemDate.day() == 5  || itemDate.day() == 6);
      }

    const isRowEditable = (record) => {
        return props.editable && (!isTotalled(record) && isWorkingDay(record) 
                              || record.isAdded || record.isUpdated
                              || isRecordUpdatedManually(record, 'entry') 
                              || hasSytemNotes(record));
    } 

    const findMaxStripId = (day) => {
        const dailyItems = tableData.filter( (item ) =>
          item.day === day
        )
        const max = dailyItems.reduce( (accu, current) => {
          return current.day === day  ? current.stripId : 0
        }, 1);

        return max;
      }

    const isRecordUpdatedManually = (record, columnName) => {

        if( !manualUpdates )
                return false;

        let found = false;  
        if( columnName === 'entry' ) {
                found = manualUpdates.find( item => {
                return item.Day == parseInt(record.day) && item.InOut === true
                })
        }  else if ( columnName === 'exit') {
                found = manualUpdates.find( item => {
                return item.Day == parseInt(record.day) && item.InOut === false
                })
        }
    
        return found ? true: false;
    }

    const onCancelAdd = () => 
        setAddModalVisible(false);

    const onCancelFullDayReport = () =>
        setFullDayReportVisible(false);        
 

    const replaceRecord = (newValues, recordKey) => {
 
        const newData = [...tableData];
        const index = newData.findIndex(item => recordKey === item.key);
        if (index > -1) {
            const item = newData[index];

            let replacedItem = {
                ...item,
                ...newValues,
                valid: true
            }

            newData.splice(index, 1, replacedItem);
            setTableData(newData);
        }

        setFullDayReportVisible(false);
        setRecordToAdd(null);        

    }

    const addRecord = ({inTime, outTime, reportCode, userNotes, isFullDay}) => {
    
        setAddModalVisible(false);
        
        let addedPositions = tableData.reduce( (accu, current, index) => {
            return {
                key: Math.max(accu.key, parseInt(current.key)),
                position: index
            }      
        },  {key:0,
            position: 0})
    
        const index = tableData.findIndex( item => 
            item.key == recordToAdd.key
            ) + 1     

        const newStripId = findMaxStripId(recordToAdd.day) + 1;

        let newItem = {
            ...recordToAdd,
            rdate: recordToAdd.rdate,
            key: addedPositions.key + 1,
            isAdded: true,
            userNotes: userNotes,
            entry: inTime,
            exit: outTime,
            stripId: newStripId,
            systemNotes: '',
            reportCode: reportCode,
            isFullDay: isFullDay
        }    
        setRecordToAdd(null);  

        newItem.total = moment.utc(moment(newItem.exit, format).diff(moment(newItem.entry, format))).format(format) 
    
        dispatch(
            action_ItemAdded(newItem, index)
        );  
        
        const newData = [
            ...tableData.slice(0, index),
            newItem,
            ...tableData.slice(index)
        ]    
    
        setTableData(newData);
    }

    const edit = (record) => {
        form.setFieldsValue({
          ...record
        })
        setEditingKey(record.key);
      }

    const editFullDay = (record) => {
        setRecordToAdd(record);
        setFullDayReportVisible(true);      
    }

    const cancel = () => {
        setEditingKey('');
    };

    const save = async (key) => {
    
        try {

          const row = await form.validateFields();
          const entryTime = row.entry;
          const exitTime = row.exit;

          if( exitTime.isBefore(entryTime) ) {
            form.setFields([{
              name: 'entry',
              value: entryTime,
              errors: [t('exit_before_entry')]
            }]);
            return;
          }

          const inouts = [entryTime, exitTime]; 

          const newData = [...tableData];
          const index = newData.findIndex(item => key === item.key);
          if (index > -1) {
            const item = newData[index];
            let newItem = {
              ...item,
              ...row,
              rdate: moment(item.rdate, 'DD/MM/YYYY').startOf('day').format('DD/MM/YYYY')
            }
            newItem.total = moment.utc(moment(newItem.exit, format).diff(moment(newItem.entry, format))).format(format)
            newItem.valid = true;
            
            newData.splice(index, 1, newItem);
            setEditingKey('');
            props.onSave && props.onSave(newItem, inouts);        
            setTableData(newData)
          }

        } catch( errorInfo ) {
          console.error(errorInfo)
        }

    }
    
    const components = {
        body: {
          cell: EditableCell,
        },
    };

    let columns = [{
        title: '',
        dataIndex: 'add',
        align: 'center',
        width: '6%',
        editable: false,
        render: (_, record) => 
          props.editable ? (
            <Row>
              <Col span={12}>
                <Tooltip title={t('add_record')}>
                  <PlusCircleTwoTone
                        onClick={() => handleAddRow(record)}/>
                </Tooltip>      
              </Col>
              <Col span={12}>
              {
                  record.isAdded ? 
                    <Popconfirm
                      title={t('sure')}
                      onConfirm={() => handleRemoveRecord(record)}>
                        <MinusCircleTwoTone />  
                    </Popconfirm>    
                  : null
              }
              </Col>
            </Row> 
            ) : null
      },
        {
        title: t('in'),
        width: '15%',
        dataIndex: 'entry',
        align: 'right',
        editable: true,
        render: (text, record) => {
            let tagColor = 'green';

            return <Tag color={tagColor}
                        style={{
                        marginRight: '0'
                    }}>
                    {
                        moment.isMoment(text) ?
                            text.format(format) : text
                    }
                    </Tag>
        }          
    }, {
        title: t('out'),
        width: '15%',
        dataIndex: 'exit',
        align: 'right',
        editable: true,
        render: (text, record) => {
            let tagColor = 'green';
            
            return <>
                <Tag color={tagColor}
                style={{
                    marginRight: '0'
                }}>
                        {
                            moment.isMoment(text) ?
                                text.format(format) : text
                        }
                </Tag>
            </>
        }
    }, {
        title: t('report_code'),
        width: '14%',
        dataIndex: 'reportCode',
        align: 'right',
        editable: true,
        render: (text, record) => {
          
          if( !moment(record.rdate, 'DD/MM/YYYY').isBefore(moment()) )
            return null;

          return <Row>
            <Col>
              <div style={{whiteSpace: 'nowrap'}}>
                {text}
              </div>
            </Col>
            <Col>
            {
              record.isFullDay ?
                manuallyEditedTag(true) : null
            }
            </Col>
          </Row>
        }
      },
      {
        title: t('system_notes'),
        dataIndex: 'systemNotes',
        align: 'right',
        editable: false,
        render: (text, record) => {

          if( !moment(record.rdate, 'DD/MM/YYYY').isBefore(moment()) )
            return null;

          return ( text !== '' ) ?
              <Tag color="blue"
                style={{
                  marginRight: '0'
                }}>
                {text}
              </Tag>
              : null
        }
      },           {
        title: t('user_notes'),
        dataIndex: 'userNotes',
        align: 'right',
        editable: true,
        render: (text, _) => {
          return text && text.length > 20 ?
                <Tooltip title={text}>
                  <Ellipsis length={12}>{text}</Ellipsis>
                </Tooltip> :
                  <div style={{
                    whiteSpace: 'nowrap'
                  }}>{text}</div>
        }
      }, {
        title: '',
        dataIndex: 'operation',
        width: '10%',
        render: (_, record) => {

          return ( moment(record.rdate, 'DD/MM/YYYY').isBefore(moment()) // no edits for future
                    && record.requireChange)? 
            (<EditIcons 
                record={record}
                editing={isRowEditing(record)} 
                disable={editingKey !== ''} 
                edit={edit}
                editFullDay={editFullDay}
                save={save} 
                cancel={cancel}
            />) : null

        }
      },
    ];

    columns = columns.map(col => {
        if (!col.editable) 
            return col;

        return {
            ...col,
            onCell: (record, rowIndex) => {
        
                return {
                record,
                inputType: getInputType(col.dataIndex),
                dataIndex: col.dataIndex,
                title: col.title,
                rowEditing: isRowEditing(record),
            }}
        };            
    })

    const getInputType = (type) => {
        const controls = {
          entry: function () {
            return 'time';
          },
          exit: function () {
            return 'time';
          },
          reportCode: function () {
            return 'select';
          },
          default: function () {
            return 'text';
          }
        };
        return (controls[type] || controls['default'])();
    }     

    return <ReportContext.Provider value={ {
        codes: reportCodes
       }
     }>
        
        <AddRecordModal 
            visible={addModalVisible}
            record = {recordToAdd}
            onCancel={onCancelAdd}
            onAddRecord={addRecord}
        />

        <Suspense fallback={<div>Loading FullDayReport...</div>}>
            <FullDayReport visible={fullDayReportVisible}
                          onCancel={onCancelFullDayReport}
                          onOk={replaceRecord}
                          record={recordToAdd}
                    />
        </Suspense>

        <Form form={form} component={false}>
            <Table style={{
                    marginRight: "38px",
                    marginLeft: "8px"
                }}
                pagination={false}
                dataSource={tableData}
                columns={columns}
                components={components}
                tableLayout='auto'
                bordered={false}  
                rowClassName="editable-row"
                size="small"
            />
        </Form>
    </ReportContext.Provider>
}

export default DailyTable;