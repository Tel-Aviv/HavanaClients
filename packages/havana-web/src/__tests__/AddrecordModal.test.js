import 'babel-polyfill'

import React from 'react';
import renderer from 'react-test-renderer';
//import {cleanup, fireEvent, render} from '@testing-library/react';

import i18n from 'i18next';
import { useTranslation, initReactI18next } from "react-i18next";
import translations from '../translations';

import { ReportContext } from "../components/reports//TableContext";
import AddRecordModal from '../components/reports/AddRecordModal';

i18n
  .use(initReactI18next) // passes i18n down to react-i18next
  .init({
    resources: translations,
    lng: "he",
    debug: false
  })

const year = 2020;
const month = 7; 
let reportCodes = [];

describe('Enables adding new entry to the report', () => {

    const OLD_ENV = process.env;

    beforeAll( async() => {
        //
        // Set 'mock' environment variable 
        // and load API dynamically that should be mocked
        //
        jest.resetModules() //clears the cache
        process.env = { ...OLD_ENV, 
                        mock: true }; // make a copy
        // Dynamically import API with 'mock' environemnt variable
        const API = require('../utils').API;

        let resp = await Promise.all([
            API.get('/me', { withCredentials: true }),
            API.get(`/me/reports/${year}/${month}`, { withCredentials: true })
        ])

        const ID = resp[0].data.ID;

        const report = resp[1].data;
        const employerCode = report.employerCode || 0;

        resp = await API.get(`/me/report_codes`, {
            params: {
                id : ID,
                employerCode: employerCode,
                year: year,
                month: month
            },
            withCredentials: true
        })
    })

    afterAll(() => {
        // Restore environment
        process.env = OLD_ENV; // restore old env
    });

    test('Fit Layout', () => {
        const component = renderer.create(
            <ReportContext.Provider value={ {
                                                codes: reportCodes
                                            }
                                           }>
                <AddRecordModal />
            </ReportContext.Provider>
        );

        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();

        //console.log(tree.props)
    })
})