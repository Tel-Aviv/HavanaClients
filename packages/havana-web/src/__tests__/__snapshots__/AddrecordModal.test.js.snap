// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Enables adding new entry to the report Fit Layout 1`] = `null`;
